web/controllers/userController/index.ts

```diff
@@ -1007,7 +1011,66 @@ async function putActive(wix_id: string, active: boolean, person: Pupil | Studen
                 await dissolveMatch(matches[i], 0, person);
             }
 
-            // Step 2: Deactivate
+            // Step 2: Cancel all courses if user is student
+            if (type == Student) {
+                let courses = await getConnection()
+                    .getRepository(Course)
+                    .createQueryBuilder("course")
+                    .leftJoinAndSelect("course.instructors", "instructors")
+                    .getMany();
+
+                courses.forEach(async (course: Course) => {
+                    if (course.instructors.some(i => i.id === person.id)) { // Only proceed if we're part of this course as an instructor
+                        if (course.instructors.length > 1) {
+                            // Course still has other instructors, only remove our person from those. We don't want to cancel those courses.
+                            course.instructors = course.instructors.filter(s => s.id !== person.id);
+                            entityManager.transaction(async em => {
+                                await em.save(Course, course);
+                                logger.info("Removed instructor " + person.firstname + " " + person.lastname + " from course " + course.name + ".");
+                            });
+                        } else {
+                        // Our person is the only instructor in the course. Cancel it.
+
+                            // We have a non-mitigated race condition here: Someone could post a new subcourse into the course, while the course gets cancelled
+                            try {
+                            // Run in transaction, so we may not have a mixed state, where some subcourses are cancelled, but others are not
+                                await entityManager.transaction(async em => {
+                                    if (course.hasOwnProperty("subcourses")) {
```

Comment:
Are there cases when the subcourses property is not set at all? Shouldn't it be an empty array then?

Reply:
Oh yeah, it's not nullable.